lib/liquid/context.rb

```diff
@@ -203,8 +229,39 @@ def lookup_and_evaluate(obj, key, raise_on_not_found: true)
       end
     end
 
+    protected
+
+    attr_writer :base_scope_depth, :warnings, :errors
+
     private
 
+    attr_reader :base_scope_depth
+
+    # As a legacy quirk, counters (the values modified by the `increment` and
+    # `decrement` tags) are stored in the highest environment, not as
+    # variables in the current scope.
+    #
+    # Since environments are shared between isoated scopes, this would lead
+    # to a state leak and prevent us from caching template fragments like those
+    # used by `{% render %}`. We still want to inherit the rest of the environment,
+    # so we track which values come from counters and reset those when creating
+    # a subscope.
+    def environments_with_reset_counters
+      cleansed_counter_environment = environments.first.clone
+      registers.fetch(:counter_names) { Set.new }.each do |counter_name|
```

Comment:
Is this equivalent to `registers[:counter_names]&.each do |counter_name|`?

Reply:
Yep! It would be. Not sure that lonely operator is supported in all our Ruby versions, though.